Assets/HoloToolkit/SpatialMapping/Scripts/SpatialMappingSource.cs

```diff
@@ -70,77 +103,189 @@ protected GameObject AddSurfaceObject(Mesh mesh, string objectName, Transform pa
+
+                    removed = surface;
                     break;
                 }
             }
+
+            return removed;
+        }
+
+        /// <summary>
+        /// Clean up the resources associated with the surface.
+        /// </summary>
+        /// <param name="surface">The surface whose resources will be cleaned up.</param>
+        /// <param name="destroyGameObject"></param>
+        /// <param name="destroyMeshes"></param>
+        /// <param name="objectToPreserve">If the surface's game object matches this parameter, it will not be destroyed.</param>
+        /// <param name="meshToPreserveA">If either of the surface's meshes matches this parameter, it will not be destroyed.</param>
+        /// <param name="meshToPreserveB">If either of the surface's meshes matches this parameter, it will not be destroyed.</param>
+        protected void CleanUpSurface(
+            SurfaceObject surface,
+            bool destroyGameObject = true,
+            bool destroyMeshes = true,
+            GameObject objectToPreserve = null,
+            Mesh meshToPreserveA = null,
+            Mesh meshToPreserveB = null
+            )
+        {
+            if (destroyGameObject
+                && (surface.Object != null)
+                && (surface.Object != objectToPreserve)
+                )
+            {
+                Destroy(surface.Object);
+                Debug.Assert(surface.GetType().IsValueType(), "If surface is no longer a value type, you should probably set surface.Object to null.");
+            }
+
+
+            Mesh filterMesh = surface.Filter.sharedMesh;
+            Mesh colliderMesh = surface.Collider.sharedMesh;
+
+            if (destroyMeshes
+                && (filterMesh != null)
+                && (filterMesh != meshToPreserveA)
+                && (filterMesh != meshToPreserveB)
+                )
+            {
+                Destroy(filterMesh);
+                surface.Filter.sharedMesh = null;
+            }
+
+            if (destroyMeshes
+                && (colliderMesh != null)
+                && (colliderMesh != filterMesh)
+                && (colliderMesh != meshToPreserveA)
+                && (colliderMesh != meshToPreserveB)
+                )
+            {
+                Destroy(colliderMesh);
+                surface.Collider.sharedMesh = null;
+            }
         }
 
         /// <summary>
         /// Cleans up references to objects that we have created.
         /// </summary>
-        protected void Cleanup()
+        /// <param name="destroyGameObjects">True to destroy the game objects of each surface, false otherwise.</param>
+        /// <param name="destroyMeshes">True to destroy the meshes of each surface, false otherwise.</param>
+        protected void Cleanup(bool destroyGameObjects = true, bool destroyMeshes = true)
         {
-            for (int index = 0; index < SurfaceObjects.Count; index++)
+            var handlers = RemovingAllSurfaces;
```

Comment:
This one I'd like to keep as "var".  Same rationale as the other "handlers" one.